File: core/api_key_vault.py
import os
from typing import Dict, List, Optional, Set
from dotenv import load_dotenv, find_dotenv, set_key, unset_key
from core.types_registry import NodeRegistry, SerialisableGraph

class APIKeyVault:
    """Singleton class for managing API keys stored in environment variables."""

    _instance = None
    _keys: Dict[str, str] = {}

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            # Load environment variables from resolved .env, allowing file to override existing env vars
            load_dotenv(cls._resolve_dotenv_path(), override=True)
            # Cache all keys that start with typical API key prefixes or known keys
            for key, value in os.environ.items():
                if any(prefix in key.upper() for prefix in ['API', 'KEY', 'TOKEN', 'SECRET']) or \
                   key in ['POLYGON_API_KEY', 'TAVILY_API_KEY', 'OLLAMA_API_KEY']:
                    cls._keys[key] = value
        return cls._instance

    def get(self, key: str) -> Optional[str]:
        """Get an API key by name."""
        return self._keys.get(key) or os.getenv(key)

    def set(self, key: str, value: str) -> None:
        """Set an API key and persist it to the .env file."""
        self._keys[key] = value
        dotenv_path = self._resolve_dotenv_path()
        if not dotenv_path:
            # Create .env file relative to cwd only when no test override provided
            dotenv_path = '.env'
        set_key(dotenv_path, key, value)
        # Update environment variable
        os.environ[key] = value

    def get_all(self) -> Dict[str, str]:
        """Get all API keys."""
        # Return a copy of cached keys plus any new ones from environment
        result = self._keys.copy()
        for key, value in os.environ.items():
            if any(prefix in key.upper() for prefix in ['API', 'KEY', 'TOKEN', 'SECRET']) or \
               key in ['POLYGON_API_KEY', 'TAVILY_API_KEY', 'OLLAMA_API_KEY']:
                result[key] = value
        return result
    
    def get_required_for_graph(self, graph: SerialisableGraph, node_registry: NodeRegistry) -> List[str]:
        """Get all required API keys for a given graph.

        Args:
            graph: The serializable graph to analyze
            node_registry: Optional registry of node types. If None, will attempt to import NODE_REGISTRY.
        
        This resolves node classes dynamically from the provided node registry
        to avoid hardcoding specific node types.
        """
        
        required_keys: Set[str] = set()

        for node_data in graph.get('nodes', []):
            node_type = node_data.get('type', '')
            cls = node_registry.get(node_type)
            if not cls:
                continue
            keys = getattr(cls, 'required_keys', []) or []
            for key in keys:
                if isinstance(key, str) and key:
                    required_keys.add(key)
        return list(required_keys)

    def unset(self, key: str) -> None:
        """Remove an API key from cache, environment, and .env file."""
        if key in self._keys:
            del self._keys[key]
        if key in os.environ:
            del os.environ[key]
        dotenv_path = self._resolve_dotenv_path()
        if dotenv_path:
            unset_key(dotenv_path, key)
        # No error if key didn't exist or .env missing - idempotent

    @staticmethod
    def _resolve_dotenv_path() -> str:
        """Find the .env file using standard dotenv resolution."""
        return find_dotenv()

File: packages/core/src/core/api-key-vault.ts
// backend/core/api-key-vault.ts
// Translated from: core/api_key_vault.py

import * as fs from 'fs';
import * as path from 'path';
import * as dotenv from 'dotenv';
import type { NodeRegistry, SerialisableGraph } from './types';

/**
 * Singleton class for managing API keys stored in environment variables.
 */
export class APIKeyVault {
  private static instance: APIKeyVault | null = null;
  private keys: Map<string, string> = new Map();
  private dotenvPath: string;

  private constructor() {
    this.dotenvPath = this.resolveDotenvPath();
    this.loadKeys();
  }

  /**
   * Get the singleton instance.
   */
  static getInstance(): APIKeyVault {
    if (!APIKeyVault.instance) {
      APIKeyVault.instance = new APIKeyVault();
    }
    return APIKeyVault.instance;
  }

  /**
   * Reset the singleton instance (useful for testing).
   */
  static reset(): void {
    APIKeyVault.instance = null;
  }

  /**
   * Load environment variables and cache relevant keys.
   */
  private loadKeys(): void {
    // Load .env file
    if (this.dotenvPath && fs.existsSync(this.dotenvPath)) {
      dotenv.config({ path: this.dotenvPath, override: true });
    }

    // Cache all keys that start with typical API key prefixes
    const prefixes = ['API', 'KEY', 'TOKEN', 'SECRET'];
    const knownKeys = ['POLYGON_API_KEY', 'TAVILY_API_KEY', 'OLLAMA_API_KEY', 'OPENROUTER_API_KEY'];

    for (const [key, value] of Object.entries(process.env)) {
      if (value === undefined) continue;

      const upperKey = key.toUpperCase();
      const hasPrefix = prefixes.some((prefix) => upperKey.includes(prefix));
      const isKnownKey = knownKeys.includes(key);

      if (hasPrefix || isKnownKey) {
        this.keys.set(key, value);
      }
    }
  }

  /**
   * Get an API key by name.
   */
  get(key: string): string | undefined {
    return this.keys.get(key) ?? process.env[key];
  }

  /**
   * Set an API key and persist it to the .env file.
   */
  set(key: string, value: string): void {
    this.keys.set(key, value);
    process.env[key] = value;

    // Persist to .env file
    const envPath = this.dotenvPath || '.env';
    this.updateEnvFile(envPath, key, value);
  }

  /**
   * Get all API keys.
   */
  getAll(): Record<string, string> {
    const result: Record<string, string> = {};

    // Include cached keys
    for (const [key, value] of this.keys) {
      result[key] = value;
    }

    // Include any new keys from environment
    const prefixes = ['API', 'KEY', 'TOKEN', 'SECRET'];
    const knownKeys = ['POLYGON_API_KEY', 'TAVILY_API_KEY', 'OLLAMA_API_KEY', 'OPENROUTER_API_KEY'];

    for (const [key, value] of Object.entries(process.env)) {
      if (value === undefined) continue;

      const upperKey = key.toUpperCase();
      const hasPrefix = prefixes.some((prefix) => upperKey.includes(prefix));
      const isKnownKey = knownKeys.includes(key);

      if (hasPrefix || isKnownKey) {
        result[key] = value;
      }
    }

    return result;
  }

  /**
   * Get all required API keys for a given graph.
   */
  getRequiredForGraph(graph: SerialisableGraph, nodeRegistry: NodeRegistry): string[] {
    const requiredKeys = new Set<string>();

    const nodes = graph.nodes ?? [];
    for (const nodeData of nodes) {
      const nodeType = nodeData.type;
      const NodeClass = nodeRegistry[nodeType];

      if (!NodeClass) {
        continue;
      }

      // Check for required_keys static property on the node class
      const keys = (NodeClass as unknown as { required_keys?: string[] }).required_keys ?? [];
      for (const key of keys) {
        if (typeof key === 'string' && key) {
          requiredKeys.add(key);
        }
      }
    }

    return Array.from(requiredKeys);
  }

  /**
   * Remove an API key from cache, environment, and .env file.
   */
  unset(key: string): void {
    this.keys.delete(key);
    delete process.env[key];

    if (this.dotenvPath && fs.existsSync(this.dotenvPath)) {
      this.removeFromEnvFile(this.dotenvPath, key);
    }
  }

  /**
   * Find the .env file using standard resolution.
   */
  private resolveDotenvPath(): string {
    // Search upward from current directory
    let currentDir = process.cwd();
    const root = path.parse(currentDir).root;

    while (currentDir !== root) {
      const envPath = path.join(currentDir, '.env');
      if (fs.existsSync(envPath)) {
        return envPath;
      }
      currentDir = path.dirname(currentDir);
    }

    // Default to .env in current directory
    return path.join(process.cwd(), '.env');
  }

  /**
   * Escape special regex characters in a string.
   */
  private escapeRegex(str: string): string {
    return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  }

  /**
   * Quote a value for .env file if it contains special characters.
   */
  private quoteEnvValue(value: string): string {
    // If value contains spaces, quotes, or newlines, wrap in double quotes and escape internal quotes
    if (/[\s"'\\]/.test(value) || value.includes('\n')) {
      return `"${value.replace(/\\/g, '\\\\').replace(/"/g, '\\"')}"`;
    }
    return value;
  }

  /**
   * Update or add a key in the .env file.
   */
  private updateEnvFile(filePath: string, key: string, value: string): void {
    let content = '';

    if (fs.existsSync(filePath)) {
      content = fs.readFileSync(filePath, 'utf-8');
    }

    const lines = content.split('\n');
    const keyRegex = new RegExp(`^${this.escapeRegex(key)}=`);
    const quotedValue = this.quoteEnvValue(value);
    let found = false;

    const newLines = lines.map((line) => {
      if (keyRegex.test(line)) {
        found = true;
        return `${key}=${quotedValue}`;
      }
      return line;
    });

    if (!found) {
      newLines.push(`${key}=${quotedValue}`);
    }

    fs.writeFileSync(filePath, newLines.join('\n'));
  }

  /**
   * Remove a key from the .env file.
   */
  private removeFromEnvFile(filePath: string, key: string): void {
    if (!fs.existsSync(filePath)) {
      return;
    }

    const content = fs.readFileSync(filePath, 'utf-8');
    const lines = content.split('\n');
    const keyRegex = new RegExp(`^${this.escapeRegex(key)}=`);

    const newLines = lines.filter((line) => !keyRegex.test(line));

    fs.writeFileSync(filePath, newLines.join('\n'));
  }
}

// Export singleton getter for convenience
export function getVault(): APIKeyVault {
  return APIKeyVault.getInstance();
}
